src/QsCompiler/SyntaxProcessor/CapabilityInference/ConstAnalyzer.fs

```diff
@@ -0,0 +1,181 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+module Microsoft.Quantum.QsCompiler.SyntaxProcessing.CapabilityInference.ConstAnalyzer
+
+open Microsoft.Quantum.QsCompiler
+open Microsoft.Quantum.QsCompiler.DataTypes
+open Microsoft.Quantum.QsCompiler.Diagnostics
+open Microsoft.Quantum.QsCompiler.SyntaxProcessing.CapabilityInference
+open Microsoft.Quantum.QsCompiler.SyntaxProcessing.CapabilityInference.ContextRef
+open Microsoft.Quantum.QsCompiler.SyntaxTokens
+open Microsoft.Quantum.QsCompiler.SyntaxTree
+open Microsoft.Quantum.QsCompiler.Transformations.Core
+
+type Context = { IsEntryPoint: bool; ConstOnly: bool }
+
+let createPattern range =
+    let capability = TargetCapability.withClassical ClassicalCapability.full TargetCapability.bottom
+
+    let diagnose (target: Target) =
+        let range = QsNullable.defaultValue Range.Zero range
+
+        if TargetCapability.subsumes target.Capability capability then
+            None
+        else
+            let description = "conditional expression or mutable variable in a constant context"
```

Comment:
I am a bit worried about users not knowing what they need to do. What do you think about something very explicit like "mutable variables and conditional expressions cannot be used in a constant context"? I also think we need to somewhere define what a constant context is - maybe even spelling that out in the error message itself.

Reply:
> What do you think about something very explicit like "mutable variables and conditional expressions cannot be used in a constant context"?

The full message is:

> This construct requires a classical runtime capability that is not supported by the target {targetName}: conditional expression or mutable variable in a constant context.

You would like to change it to:

> This construct requires a classical runtime capability that is not supported by the target {targetName}: mutable variables and conditional expressions cannot be used in a constant context.

This is a bit inconsistent from the other error messages, where the description after the colon is a noun phrase describing what you did (e.g., "recursion" or "fail statement"), rather than a complete sentence. Since the first half of the message says "a ... capability that is not supported," IMO it's already clear that what is being described is not allowed?

> I also think we need to somewhere define what a constant context is - maybe even spelling that out in the error message itself.

A constant context is one of:

- The iterator in a for loop.
- The length `n` of an array in `[x, size = n]` or `Qubit[n]`.
- The value of an immutable variable binding.
- An array index.
- A call argument.

It's hard to describe all of that succinctly in an error message - I would prefer to link to documentation, but we don't have any right now.

Reply:
Alright, let's leave it as is, and we should think about documentation (not sure if we need to formalize further first).